File: src/main/java/cn/labzen/web/response/format/UnexpectedResponseFormatter.java
package cn.labzen.web.response.format;

import cn.labzen.web.response.bean.Meta;
import cn.labzen.web.response.bean.Response;
import cn.labzen.web.response.bean.Result;
import com.google.common.primitives.Longs;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import java.util.Optional;

import static cn.labzen.web.defination.Constants.REST_REQUEST_TIME;
import static cn.labzen.web.defination.Constants.REST_REQUEST_TIME_MILLIS;

/**
 * 格式化非 {@link Result} 类型的返回值
 */
public class UnexpectedResponseFormatter implements ResponseFormatter {

  @Override
  public boolean support(Class<?> clazz, HttpServletRequest request) {
    return true;
  }

  @Override
  public Object format(Object result, HttpServletRequest request, HttpServletResponse response) {
    String requestTime = request.getAttribute(REST_REQUEST_TIME).toString();
    String requestMillsStr = request.getAttribute(REST_REQUEST_TIME_MILLIS).toString();
    long requestMills = Optional.ofNullable(Longs.tryParse(requestMillsStr)).orElse(0L);
    long executionTime = System.currentTimeMillis() - requestMills;

    Meta meta = new Meta(requestTime, executionTime, null, null, null);
    return new Response(200, "success", meta, result);
  }
}

File: src/main/kotlin/cn/labzen/web/response/format/UnexpectedResponseFormatter.kt
package cn.labzen.web.response.format

import cn.labzen.web.REST_REQUEST_TIME
import cn.labzen.web.REST_REQUEST_TIME_MILLIS
import cn.labzen.web.response.bean.Meta
import cn.labzen.web.response.bean.Response
import cn.labzen.web.response.bean.Result
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 格式化非 [Result] 类型的返回值
 */
class UnexpectedResponseFormatter : ResponseFormatter {

  override fun support(clazz: Class<*>, request: HttpServletRequest): Boolean =
    true

  override fun format(result: Any?, request: HttpServletRequest, response: HttpServletResponse): Any {
    val requestTime = request.getAttribute(REST_REQUEST_TIME).toString()
    val requestMillis = request.getAttribute(REST_REQUEST_TIME_MILLIS).toString().toLong()
    val executionTime = System.currentTimeMillis() - requestMillis

    return Response(200, "success", Meta(requestTime, executionTime), result)
  }

}